File: js/src/launcher/BrowserbaseBrowserLauncher.ts
import {
  BrowserLauncher,
  type BrowserLaunchOptions,
  type LaunchedBrowser,
} from "./BrowserLauncher.js";

const DEFAULT_BROWSERBASE_BASE_URL = "https://api.browserbase.com";
const DEFAULT_BROWSERBASE_VIEWPORT = { width: 1288, height: 711 };

type BrowserbaseSession = {
  id?: string;
  connectUrl?: string;
  debuggerUrl?: string;
  debuggerFullscreenUrl?: string;
  status?: string;
};

function firstString(...values: unknown[]) {
  for (const value of values) {
    if (typeof value === "string" && value.trim()) return value.trim();
  }
  return null;
}

function firstBoolean(...values: unknown[]) {
  for (const value of values) {
    if (typeof value === "boolean") return value;
  }
  return null;
}

function objectValue(value: unknown): Record<string, unknown> {
  return value && typeof value === "object" && !Array.isArray(value)
    ? (value as Record<string, unknown>)
    : {};
}

function browserbaseUrl(base_url: string, pathname: string) {
  return new URL(pathname, `${base_url.replace(/\/$/, "")}/`).toString();
}

async function browserbaseRequest<T>({
  base_url,
  browserbase_api_key,
  method,
  pathname,
  body,
}: {
  base_url: string;
  browserbase_api_key: string;
  method: "GET" | "POST";
  pathname: string;
  body?: Record<string, unknown>;
}) {
  const response = await fetch(browserbaseUrl(base_url, pathname), {
    method,
    headers: {
      "content-type": "application/json",
      "x-bb-api-key": browserbase_api_key,
    },
    body: body == null ? undefined : JSON.stringify(body),
  });
  if (!response.ok) {
    const text = await response.text().catch(() => "");
    throw new Error(
      `Browserbase ${method} ${pathname} -> ${response.status}${text ? `: ${text}` : ""}`,
    );
  }
  return (await response.json()) as T;
}

async function closeBrowserCDP(cdp_url: string | undefined) {
  if (
    !cdp_url ||
    !/^wss?:\/\//i.test(cdp_url) ||
    typeof WebSocket !== "function"
  )
    return;
  await new Promise<void>((resolve) => {
    let settled = false;
    const ws = new WebSocket(cdp_url);
    const finish = () => {
      if (settled) return;
      settled = true;
      clearTimeout(timeout);
      try {
        ws.close();
      } catch {}
      resolve();
    };
    const timeout = setTimeout(finish, 2_000);
    ws.addEventListener("open", () => {
      try {
        ws.send(JSON.stringify({ id: 1, method: "Browser.close", params: {} }));
      } catch {
        finish();
      }
    });
    ws.addEventListener("message", finish, { once: true });
    ws.addEventListener("close", finish, { once: true });
    ws.addEventListener("error", finish, { once: true });
  });
}

export class BrowserbaseBrowserLauncher extends BrowserLauncher {
  async launch(options: BrowserLaunchOptions = {}): Promise<LaunchedBrowser> {
    const merged = { ...this.options, ...options };
    const browserbase_api_key = firstString(
      merged.browserbase_api_key,
      process.env.BROWSERBASE_API_KEY,
    );
    if (!browserbase_api_key) {
      throw new Error(
        "launcher.launcher_mode=bb requires BROWSERBASE_API_KEY or launcher.launcher_options.browserbase_api_key.",
      );
    }

    const project_id = firstString(
      merged.browserbase_project_id,
      process.env.BROWSERBASE_PROJECT_ID,
    );
    const base_url =
      firstString(
        merged.browserbase_base_url,
        process.env.BROWSERBASE_BASE_URL,
      ) ?? DEFAULT_BROWSERBASE_BASE_URL;
    const resume_session_id = firstString(merged.browserbase_session_id);
    const keep_alive = firstBoolean(merged.browserbase_keep_alive) ?? false;
    const close_session_on_close =
      firstBoolean(merged.browserbase_close_session_on_close) ?? !keep_alive;

    let created_session = false;
    let session: BrowserbaseSession;
    if (resume_session_id) {
      session = await browserbaseRequest<BrowserbaseSession>({
        base_url,
        browserbase_api_key,
        method: "GET",
        pathname: `/v1/sessions/${resume_session_id}`,
      });
    } else {
      const session_create_params = objectValue(
        merged.browserbase_session_create_params,
      );
      const browser_settings = {
        ...objectValue(session_create_params.browserSettings),
        ...objectValue(merged.browserbase_browser_settings),
      };
      const user_metadata = {
        ...objectValue(session_create_params.userMetadata),
        ...objectValue(merged.browserbase_user_metadata),
      };
      const extension_id = firstString(
        merged.injector_extension_id,
        session_create_params.extensionId,
        objectValue(session_create_params.browserSettings).extensionId,
      );
      const body = {
        ...session_create_params,
        ...(project_id ? { projectId: project_id } : {}),
        ...(keep_alive ? { keepAlive: true } : {}),
        ...(firstString(merged.region, session_create_params.region)
          ? { region: firstString(merged.region, session_create_params.region) }
          : {}),
        ...(typeof merged.timeout === "number"
          ? { timeout: merged.timeout }
          : {}),
        ...(extension_id ? { extensionId: extension_id } : {}),
        browserSettings: {
          ...browser_settings,
          ...(extension_id ? { extensionId: extension_id } : {}),
          viewport: objectValue(browser_settings.viewport).width
            ? browser_settings.viewport
            : DEFAULT_BROWSERBASE_VIEWPORT,
        },
        userMetadata: {
          ...user_metadata,
          modcdp: "true",
        },
      };
      session = await browserbaseRequest<BrowserbaseSession>({
        base_url,
        browserbase_api_key,
        method: "POST",
        pathname: "/v1/sessions",
        body,
      });
      created_session = true;
    }

    if (!session.id || !session.connectUrl) {
      throw new Error(
        "Browserbase session creation returned an unexpected shape.",
      );
    }

    let closed = false;
    const close = async () => {
      if (closed) return;
      closed = true;
      if (!created_session || !close_session_on_close) return;
      await closeBrowserCDP(session.connectUrl).catch(() => {});
      await browserbaseRequest({
        base_url,
        browserbase_api_key,
        method: "POST",
        pathname: `/v1/sessions/${session.id}`,
        body: {
          status: "REQUEST_RELEASE",
          ...(project_id ? { projectId: project_id } : {}),
        },
      }).catch(() => {});
    };

    this.launched = {
      // Browserbase connectUrl is already a WebSocket CDP endpoint.
      cdp_url: session.connectUrl,
      browserbase_session_id: session.id,
      browserbase_session_url: `https://www.browserbase.com/sessions/${session.id}`,
      browserbase_debug_url:
        session.debuggerUrl ?? session.debuggerFullscreenUrl ?? null,
      close,
    };
    return this.launched;
  }
}

File: go/modcdp/launcher/BrowserbaseBrowserLauncher.go
package launcher

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"strings"
	"time"

	"github.com/gobwas/ws"
	"github.com/gobwas/ws/wsutil"
)

const DefaultBrowserbaseLauncherBaseURL = "https://api.browserbase.com"

var defaultBrowserbaseViewport = map[string]any{"width": 1288, "height": 711}

type BrowserbaseBrowserLauncher struct {
	BrowserLauncher
}

type browserbaseSession struct {
	ID                    string `json:"id"`
	ConnectURL            string `json:"connectUrl"`
	DebuggerURL           string `json:"debuggerUrl"`
	DebuggerFullscreenURL string `json:"debuggerFullscreenUrl"`
	Status                string `json:"status"`
}

func NewBrowserbaseBrowserLauncher(options LaunchOptions) *BrowserbaseBrowserLauncher {
	return &BrowserbaseBrowserLauncher{BrowserLauncher: NewBrowserLauncher(options)}
}

func (l *BrowserbaseBrowserLauncher) Launch(options LaunchOptions) (*LaunchedBrowser, error) {
	merged := mergeLaunchOptions(l.Options, options)
	browserbaseAPIKey := firstString(merged.BrowserbaseAPIKey, os.Getenv("BROWSERBASE_API_KEY"))
	if browserbaseAPIKey == "" {
		return nil, fmt.Errorf("launch.mode=bb requires BROWSERBASE_API_KEY or launch.options.browserbase_api_key")
	}

	projectID := firstString(
		merged.BrowserbaseProjectID,
		os.Getenv("BROWSERBASE_PROJECT_ID"),
	)
	baseURL := firstString(merged.BrowserbaseBaseURL, os.Getenv("BROWSERBASE_BASE_URL"), DefaultBrowserbaseLauncherBaseURL)
	resumeSessionID := firstString(merged.BrowserbaseSessionID)
	keepAlive := boolValue(merged.BrowserbaseKeepAlive, false)
	closeSessionOnClose := boolValue(merged.BrowserbaseCloseSessionOnClose, !keepAlive)

	createdSession := false
	var session browserbaseSession
	var err error
	if resumeSessionID != "" {
		err = browserbaseRequest(baseURL, browserbaseAPIKey, http.MethodGet, "/v1/sessions/"+resumeSessionID, nil, &session)
	} else {
		sessionCreateParams := objectValue(merged.BrowserbaseSessionCreateParams)
		browserSettings := mergeMap(objectValue(sessionCreateParams["browserSettings"]), objectValue(merged.BrowserbaseBrowserSettings))
		userMetadata := mergeMap(objectValue(sessionCreateParams["userMetadata"]), objectValue(merged.BrowserbaseUserMetadata))
		extensionID := firstString(
			merged.ExtensionID,
			stringValue(sessionCreateParams["extensionId"]),
			stringValue(objectValue(sessionCreateParams["browserSettings"])["extensionId"]),
		)
		body := mergeMap(map[string]any{}, sessionCreateParams)
		if projectID != "" {
			body["projectId"] = projectID
		}
		if keepAlive {
			body["keepAlive"] = true
		}
		if region := firstString(merged.Region, stringValue(sessionCreateParams["region"])); region != "" {
			body["region"] = region
		}
		if merged.Timeout != 0 {
			body["timeout"] = merged.Timeout
		}
		if extensionID != "" {
			body["extensionId"] = extensionID
			browserSettings["extensionId"] = extensionID
		}
		if !hasViewportWidth(objectValue(browserSettings["viewport"])) {
			browserSettings["viewport"] = defaultBrowserbaseViewport
		}
		userMetadata["modcdp"] = "true"
		body["browserSettings"] = browserSettings
		body["userMetadata"] = userMetadata
		err = browserbaseRequest(baseURL, browserbaseAPIKey, http.MethodPost, "/v1/sessions", body, &session)
		createdSession = true
	}
	if err != nil {
		return nil, err
	}
	if session.ID == "" || session.ConnectURL == "" {
		return nil, fmt.Errorf("Browserbase session creation returned an unexpected shape")
	}

	closed := false
	close := func() {
		if closed {
			return
		}
		closed = true
		if !createdSession || !closeSessionOnClose {
			return
		}
		closeBrowserCDP(session.ConnectURL)
		body := map[string]any{"status": "REQUEST_RELEASE"}
		if projectID != "" {
			body["projectId"] = projectID
		}
		var ignored map[string]any
		_ = browserbaseRequest(baseURL, browserbaseAPIKey, http.MethodPost, "/v1/sessions/"+session.ID, body, &ignored)
	}

	launched := &LaunchedBrowser{
		// Browserbase ConnectURL is already a WebSocket CDP endpoint.
		CDPURL:                session.ConnectURL,
		BrowserbaseSessionID:  session.ID,
		BrowserbaseSessionURL: "https://www.browserbase.com/sessions/" + session.ID,
		BrowserbaseDebugURL:   firstString(session.DebuggerURL, session.DebuggerFullscreenURL),
		Close:                 close,
	}
	l.Launched = launched
	return launched, nil
}

func browserbaseRequest(baseURL string, browserbaseAPIKey string, method string, pathname string, body map[string]any, out any) error {
	var reader io.Reader
	if body != nil {
		bodyBytes, err := json.Marshal(body)
		if err != nil {
			return err
		}
		reader = bytes.NewReader(bodyBytes)
	}
	request, err := http.NewRequest(method, strings.TrimRight(baseURL, "/")+"/"+strings.TrimLeft(pathname, "/"), reader)
	if err != nil {
		return err
	}
	request.Header.Set("content-type", "application/json")
	request.Header.Set("x-bb-api-key", browserbaseAPIKey)
	response, err := http.DefaultClient.Do(request)
	if err != nil {
		return err
	}
	defer response.Body.Close()
	responseBody, _ := io.ReadAll(response.Body)
	if response.StatusCode < 200 || response.StatusCode >= 300 {
		return fmt.Errorf("Browserbase %s %s -> %d%s", method, pathname, response.StatusCode, errorText(responseBody))
	}
	if out == nil || len(responseBody) == 0 {
		return nil
	}
	return json.Unmarshal(responseBody, out)
}

func closeBrowserCDP(wsURL string) {
	if !strings.HasPrefix(wsURL, "ws://") && !strings.HasPrefix(wsURL, "wss://") {
		return
	}
	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
	defer cancel()
	conn, _, _, err := ws.Dial(ctx, wsURL)
	if err != nil {
		return
	}
	defer conn.Close()
	_ = wsutil.WriteClientMessage(conn, ws.OpText, []byte(`{"id":1,"method":"Browser.close","params":{}}`))
	_ = conn.SetDeadline(time.Now().Add(500 * time.Millisecond))
	_, _, _ = wsutil.ReadServerData(conn)
}

func errorText(body []byte) string {
	if len(body) == 0 {
		return ""
	}
	return ": " + string(body)
}

func boolValue(value *bool, fallback bool) bool {
	if value == nil {
		return fallback
	}
	return *value
}

func firstMap(values ...map[string]any) map[string]any {
	for _, value := range values {
		if value != nil {
			return value
		}
	}
	return nil
}

func objectValue(value any) map[string]any {
	if object, ok := value.(map[string]any); ok {
		return object
	}
	return map[string]any{}
}

func mergeMap(left map[string]any, right map[string]any) map[string]any {
	merged := map[string]any{}
	for key, value := range left {
		merged[key] = value
	}
	for key, value := range right {
		merged[key] = value
	}
	return merged
}

func stringValue(value any) string {
	if value, ok := value.(string); ok {
		return strings.TrimSpace(value)
	}
	return ""
}

func hasViewportWidth(viewport map[string]any) bool {
	if _, ok := viewport["width"]; ok {
		return true
	}
	return false
}
